### DISCUSSION TITLE
Webpack 5 is tree shaking all modules required using require.resolve

### DISCUSSION THREAD

**Discussion by sachetsharma1** [webpack/webpack#18012] [Q&A]
# Bug report

**What is the current behavior?**

Webpack marks all the modules as  **unused harmony export** which gets tree shaken on the prod bundle

**If the current behavior is a bug, please provide the steps to reproduce.**

Most of the code is taken from this repo for reproduction - https://github.com/bradtraversy/webpack-starter/tree/main/src

**Changes Made**

- Added a **testMethod.js** file 
```
export default function testMethod() {
  console.log("test method called");
}

```

- Added a **registry.js** file with following content
```
export default {
  testMethodModule: require.resolve("./testMethod"),
};
```
- Added following code to **generateJoke.js**

```
import registry from "./registry";
  
...
const testMethod = __webpack_require__(registry.testMethodModule);
  console.log(testMethod);
```

- Modified webpack.config.js to the following 

```
mode: "production",
  entry: {
    bundle: path.resolve(__dirname, "src/index.js"),
  },
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "[name][contenthash].js",
    clean: true,
    assetModuleFilename: "[name][ext]",
  },
  devtool: "source-map",
  devServer: {
    static: {
      directory: path.resolve(__dirname, "dist"),
    },
    port: 4000,
    open: true,
    hot: true,
    compress: true,
    historyApiFallback: true,
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: ["@babel/preset-env"],
          },
        },
      },
    ],
  },
  plugins: [
    new HtmlWebpackPlugin({
      title: "Webpack App",
      filename: "index.html",
      template: "src/template.html",
    }),
    new BundleAnalyzerPlugin(),
  ],
  optimization: {
    minimize: false,
  },
```

When I run webpack dev server `NODE_ENV=production webpack serve`, following output is generated in bundle.js

```

/***/ 735:
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

"use strict";
/* unused harmony export default */
function testMethod() {
  console.log("test method called");
}

/***/ }),
```

**What is the expected behavior?**

testMethod should not be marked as unused. This has broken in webpack 5 since the same app with webpack 4 produces the below output.

```
/***/ "./src/testMethod.js":
/*!***************************!*\
  !*** ./src/testMethod.js ***!
  \***************************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "default", function() { return testMethod; });
function testMethod() {
  console.log("test method called");
}

/***/ }),

```

We rely heavily on this feature to register modules in our library.

**Other relevant information:**
webpack version:  5.88.2
Node.js version: 20.8.0
Operating System: Mac OS ventura 13.6.2
Additional tools:

**Comment by alexander-akait** [comment]
```js
export default {
  testMethodModule: require.resolve("./testMethod"),
};
```

There is no `require.resolve` in ESM

**Reply by sachetsharma1** [reply]
We are aware of that. Is there a recommended pattern to achieve this in ESM world? We have been looking at import meta but don't have a solution yet.

**Reply by alexander-akait** [reply]
Right now - no, we consider to support `import.meta.resolve` in future, workaround now - create a cjs files and put all `require.resolve` there, then import it
